src/lib.rs

```diff
@@ -0,0 +1,1036 @@
+            0 => InstPosition::Before,
+            1 => InstPosition::After,
+            _ => unreachable!(),
+        }
+    }
+    #[inline(always)]
+    pub fn next(self) -> ProgPoint {
+        Self {
+            bits: self.bits + 1,
+        }
+    }
+    #[inline(always)]
+    pub fn prev(self) -> ProgPoint {
+        Self {
+            bits: self.bits - 1,
+        }
+    }
+    #[inline(always)]
+    pub fn to_index(self) -> u32 {
+        self.bits
+    }
+    #[inline(always)]
+    pub fn from_index(index: u32) -> Self {
+        Self { bits: index }
+    }
+}
+
+/// An instruction to insert into the program to perform some data movement.
+#[derive(Clone, Debug)]
+pub enum Edit {
+    /// Move one allocation to another. Each allocation may be a
+    /// register or a stack slot (spillslot). However, stack-to-stack
+    /// moves will never be generated.
+    ///
+    /// `to_vreg`, if defined, is useful as metadata: it indicates
+    /// that the moved value is a def of a new vreg.
+    ///
+    /// `Move` edits will be generated even if src and dst allocation
+    /// are the same if the vreg changes; this allows proper metadata
+    /// tracking even when moves are elided.
+    Move {
+        from: Allocation,
+        to: Allocation,
+        to_vreg: Option<VReg>,
+    },
+    /// Define blockparams' locations. Note that this is not typically
+    /// turned into machine code, but can be useful metadata (e.g. for
+    /// the checker).
+    BlockParams {
+        vregs: Vec<VReg>,
+        allocs: Vec<Allocation>,
+    },
+    /// Define a particular Allocation to contain a particular VReg. Useful
+    /// for the checker.
+    DefAlloc { alloc: Allocation, vreg: VReg },
+}
+
+/// A machine envrionment tells the register allocator which registers
+/// are available to allocate and what register may be used as a
+/// scratch register for each class, and some other miscellaneous info
+/// as well.
+#[derive(Clone, Debug)]
+pub struct MachineEnv {
+    /// Physical registers. Every register that might be mentioned in
+    /// any constraint must be listed here, even if it is not
+    /// allocatable under normal conditions.
+    pub regs: Vec<PReg>,
+    /// Preferred physical registers for each class. These are the
+    /// registers that will be allocated first, if free.
+    pub preferred_regs_by_class: Vec<Vec<PReg>>,
+    /// Non-preferred physical registers for each class. These are the
+    /// registers that will be allocated if a preferred register is
+    /// not available; using one of these is considered suboptimal,
+    /// but still better than spilling.
+    pub non_preferred_regs_by_class: Vec<Vec<PReg>>,
+    /// One scratch register per class. This is needed to perform
+    /// moves between registers when cyclic move patterns occur. The
+    /// register should not be placed in either the preferred or
+    /// non-preferred list (i.e., it is not otherwise allocatable).
+    pub scratch_by_class: Vec<PReg>,
```

Comment:
The scratch register is currently only used for moves which are inserted between instructions. Does this mean that the scratch register is available to use within instructions? E.g. for instructions that expand to a sequence of multiple instructions. If so then this should be explicitly clarified in the doc comment.

Reply:
Yes, indeed! I'll clarify this, thanks.